### PR TITLE
Add BaseEstimatorV2

### PR DESCRIPTION
**Pull request by chriseclectic**
### Summary

Depends on #11524 

Adds BaseEstimatorV2 class

### Details and comments

BaseEstimatorV2 is taken from #11227 with the following modifications:

- Adds a `precision` attribute to BaseEstimatorV2 ~and EstimatorPub~
- Removes Options from BaseEstimatorV2
- Removes unnecessary BasePrimitiveV2 and BasePub classes

The `precision` attribute is intended to be the API element for controlling estimator precision in algorithms or applications (eg VQE). Previously this intended to be done via shots in Options, however as pointed out in review Options do not define an actual interface API since subclasses can implement them however they like. Subclasses can still implement options (and the runtime estimator will), but they are no longer part of the base API, but rather specific to how an individual estimator can be configured. This means the only API elements of an initialized EstimatorV2 for use in applications are a `run` method, and a `precision` getter/setter.

### CODE REVIEW

File: qiskit/primitives/base/base_estimator.py [hunk 1]
```diff
@@ -254,3 +325,43 @@ def parameters(self) -> tuple[ParameterView, ...]:
...
+    """Estimator base class version 2.
+
+    An estimator estimates expectation values for provided quantum circuit and
+    observable combinations.
+    """
+
+    def __init__(self, precision: float | None):
+        """
+        Args:
+            precision: a target precision of mean expectation value estimates.
```
Comment (ihincks) [inline qiskit/primitives/base/base_estimator.py]: To me this is clearer without the word "mean"

```suggestion
            precision: A target precision for expectation value estimates.
```

File: qiskit/primitives/base/base_estimator.py [hunk 2]
```diff
@@ -254,3 +325,43 @@ def parameters(self) -> tuple[ParameterView, ...]:
...
+        """
+        self._precision = precision
+
+    @property
+    def precision(self) -> float | None:
+        """The target precision for mean expectation value estimates."""
```
Comment (ihincks) [inline qiskit/primitives/base/base_estimator.py]:
```suggestion
        """The target precision for expectation value estimates."""
```

File: qiskit/primitives/base/base_estimator.py [hunk 3]
```diff
@@ -254,3 +325,43 @@ def parameters(self) -> tuple[ParameterView, ...]:
...
+    def precision(self, value: float | None):
+        self._precision = value
+
+    @abstractmethod
+    def run(self, pubs: Iterable[EstimatorPubLike]) -> Job:
+        """Estimate expectation values for each provided pub (Primitive Unified Bloc).
+
+        Args:
+            pubs: a iterable of pubslike object. Typically, list of tuple
+                ``(QuantumCircuit, observables, parameter_values)``
```
Comment (ihincks) [inline qiskit/primitives/base/base_estimator.py]:
```suggestion
            pubs: An iterable of pub-like objects, such as tuples ``(circuit, observables)`` or 
            ``(circuit, observables, parameter_values)``.
```

File: qiskit/primitives/containers/estimator_pub.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+from qiskit import QuantumCircuit
+
+from .bindings_array import BindingsArray, BindingsArrayLike
+from .observables_array import ObservablesArray, ObservablesArrayLike
+from .shape import ShapedMixin
+
+
+class EstimatorPub(ShapedMixin):
+    """Primitive Unified Bloc for Estimator.
+    Pub is composed of triple (circuit, observables, parameter_values).
```
Comment (ihincks) [inline qiskit/primitives/containers/estimator_pub.py]:
```suggestion
    """Primitive Unified Bloc for any Estimator primitive.
    
    An estimator pub is essentially a triple ``(circuit, observables, parameter_values)``.
```

File: qiskit/primitives/containers/estimator_pub.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        parameter_values: BindingsArray | None = None,
+        validate: bool = False,
+    ):
+        """Initialize an estimator pub.
+
+        Args:
+            circuit: a quantum circuit.
+            observables: an observables array.
+            parameter_values: a bindings array.
+            validate: if True, the input data is validated during initialization.
```
Comment (ihincks) [inline qiskit/primitives/containers/estimator_pub.py]:
```suggestion
            circuit: A quantum circuit.
            observables: An observables array.
            parameter_values: A bindings array, if the circuit is parametric.
            validate: Whether to validate arguments during initialization.
```

File: qiskit/primitives/containers/estimator_pub.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+        # For ShapedMixin
+        self._shape = np.broadcast_shapes(self.observables.shape, self.parameter_values.shape)
+
+        if validate:
+            self.validate()
+
+    @property
+    def observables(self) -> ObservablesArray:
+        """An observables array"""
```
Comment (ihincks) [inline qiskit/primitives/containers/estimator_pub.py]:
```suggestion
        """An observables array."""
```

File: qiskit/primitives/containers/estimator_pub.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+        return self._observables
+
+    @property
+    def parameter_values(self) -> BindingsArray:
+        """A bindings array"""
```
Comment (ihincks) [inline qiskit/primitives/containers/estimator_pub.py]:
```suggestion
        """A bindings array."""
```

File: qiskit/primitives/containers/estimator_pub.py [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+        return self._parameter_values
+
+    @classmethod
+    def coerce(cls, pub: EstimatorPubLike) -> EstimatorPub:
+        """Coerce EstimatorPubLike into EstimatorPub.
```
Comment (ihincks) [inline qiskit/primitives/containers/estimator_pub.py]:
```suggestion
        """Coerce :class:`~.EstimatorPubLike` into :class:`~.EstimatorPub`.
```

File: qiskit/primitives/containers/estimator_pub.py [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+
+        Args:
+            pub: an object to be estimator pub.
```
Comment (ihincks) [inline qiskit/primitives/containers/estimator_pub.py]:
```suggestion
            pub: A compatible object for coersion.
```

File: qiskit/primitives/containers/estimator_pub.py [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+
+        Returns:
+            A coerced estimator pub.
```
Comment (ihincks) [inline qiskit/primitives/containers/estimator_pub.py]:
```suggestion
            An estimator pub.
```

File: qiskit/primitives/base/base_estimator.py [hunk 11]
```diff
@@ -254,3 +325,43 @@ def parameters(self) -> tuple[ParameterView, ...]:
...
+
+
+class BaseEstimatorV2:
+    """Estimator base class version 2.
+
+    An estimator estimates expectation values for provided quantum circuit and
+    observable combinations.
+    """
+
+    def __init__(self, precision: float | None):
```
Comment (ihincks) [inline qiskit/primitives/base/base_estimator.py]: I'd prefer to force this into a numeric value unless there is a reason to have a None state: it would be nice to avoid `is None` checks where possible.

```suggestion
    def __init__(self, precision: float):
```

(likewise in `@property`)

File: qiskit/primitives/base/base_estimator.py [hunk 12]
```diff
@@ -1,6 +1,6 @@
 # This code is part of Qiskit.
 #
-# (C) Copyright IBM 2022.
+# (C) Copyright IBM 2022, 2023.
```
Comment (ihincks) [inline qiskit/primitives/base/base_estimator.py]:
```suggestion
# (C) Copyright IBM 2022, 2023, 2024.
```

File: qiskit/primitives/containers/estimator_pub.py [hunk 13]
```diff
@@ -0,0 +1,123 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2023.
```
Comment (ihincks) [inline qiskit/primitives/containers/estimator_pub.py]:
```suggestion
# (C) Copyright IBM 2024.
```

**Review by ihincks — COMMENTED** [review]
This review is a string of docstring touch-ups.

File: qiskit/primitives/base/base_estimator.py [hunk 14]
```diff
@@ -254,3 +328,48 @@ def parameters(self) -> tuple[ParameterView, ...]:
...
+
+
+class BaseEstimatorV2:
+    """Estimator base class version 2.
+
+    An estimator estimates expectation values for provided quantum circuit and
+    observable combinations.
+    """
+
+    def __init__(self, precision: float):
```
Comment (ikkoham) [inline qiskit/primitives/base/base_estimator.py]: I understand the intent to include precision. But, why did you not let PUB have this? In other words, given multiple PUBs, wouldn't user want to change the precision for each PUB?
Also, is it not optional? Do users ever want to set the accuracy on the shots side?

My suggestion is to make precision a specification that is implementation-dependent, rather than at the base class level.

**Review by ikkoham — COMMENTED** [review]
Thank you @chriseclectic. Just one comment (question).

**Comment by 1ucian0** [comment]
One of the `TODO`s introduced in #11524 seems to be depending on a closed issue. Can that `TODO` be addressed in this PR?

> ```
>         # TODO str will be used for internal data (_kwvals) instead of Parameter.
>         # This requires https://github.com/Qiskit/qiskit/issues/7107
> ```

File: qiskit/primitives/base/base_estimator.py [hunk 15]
```diff
@@ -254,3 +328,41 @@ def parameters(self) -> tuple[ParameterView, ...]:
...
+    """
+
+    @staticmethod
+    def _make_data_bin(pub: EstimatorPub) -> DataBin:
+        # provide a standard way to construct estimator databins to ensure that names match
+        # across implementations
+        return make_data_bin((("evs", NDArray[np.float]), ("stds", NDArray[np.float])), pub.shape)
+
+    @abstractmethod
+    def run(self, pubs: Iterable[EstimatorPubLike], precision: float | None = None) -> Job:
```
Comment (jyu00) [inline qiskit/primitives/base/base_estimator.py]: Why is `precision` not part of options, when everything else is?

File: qiskit/primitives/containers/estimator_pub.py [hunk 16]
```diff
@@ -0,0 +1,168 @@
...
+
+    __slots__ = ("_circuit", "_observables", "_parameter_values", "_precision", "_shape")
+
+    def __init__(
+        self,
+        circuit: QuantumCircuit,
+        observables: ObservablesArray,
+        parameter_values: BindingsArray | None = None,
+        precision: float | None = None,
+        validate: bool = False,
```
Comment (jyu00) [inline qiskit/primitives/containers/estimator_pub.py]: Shouldn't the default be `True`? I'd think most people prefer validation.

**Review by jyu00 — COMMENTED** [review]
What about estimator job results? The `PubResult` container doesn't really tell me what exactly data is returned.

**Comment by jyu00** [comment]
> Removes Options from BaseEstimatorV2

Why was `options` removed from the base class? I understand the desire to keep the base class simple, but the user feedback we got was that it's preferred to have some consistency (e.g. same options related attributes/methods) across different primitive implementations. This was also documented in the [RFC](https://github.com/Qiskit/RFCs/blob/master/0018-primitive-options-type.md).

File: qiskit/primitives/containers/estimator_pub.py (hunk 16 above)
Comment (chriseclectic) [reply qiskit/primitives/containers/estimator_pub.py]: Yeah I would agree we should probably default to True for validation.

File: qiskit/primitives/base/base_estimator.py (hunk 15 above)
Comment (ikkoham) [reply qiskit/primitives/base/base_estimator.py]: I also don't get the priority: if there is an option, a constructor argument, and a run argument, what has priority?

Comment (t-imamichi) [inline qiskit/primitives/base/base_estimator.py]: I suggest returning `BasePrimitiveJob` discussing at #11552.

File: qiskit/primitives/base/base_estimator.py [hunk 17]
```diff
@@ -14,9 +14,79 @@
...
-=====================
+========================
+Overview of EstimatorV2
+========================
+
+:class:`~BaseEstimatorV2` is a primitive that estimates expectation values for provided quantum
+circuit and observable combinations.
+
+Following construction, an estimator is used by calling its :meth:`~.BaseEstimatorV2.run` method
+with a list of pubs (Primitive Unified Blocs). Each pub contains three values that, together,
```
Comment (ihincks) [inline qiskit/primitives/base/base_estimator.py:25]:
```suggestion
with a list of pubs (Primitive Unified Blocs). Each pub contains values that, together,
```

File: qiskit/primitives/base/base_estimator.py [hunk 18]
```diff
@@ -14,9 +14,79 @@
...
+define a computation unit of work for the estimator to complete:
+
+* a single :class:`~qiskit.circuit.QuantumCircuit`, possibly parametrized, whose final state we
+  define as :math:`\psi(\theta)`,
+
+* one or more observables (specified as any :class:`~.ObservablesArrayLike`, including
+  :class:`~.Pauli`, :class:`~.SparsePauliOp`, ``str``) that specify which expectation values to
+  estimate, denoted :math:`H_j`, and
+
+* a collection parameter value sets to bind the circuit against, :math:`\theta_k`.
```
Comment (ihincks) [inline qiskit/primitives/base/base_estimator.py:35]:
```suggestion
  estimate, denoted :math:`H_j`,

* a collection parameter value sets to bind the circuit against, :math:`\theta_k`, and

* optionally, a target precision `\epsilon` to meet or exceed for each estimate.
```

File: qiskit/primitives/base/base_estimator.py [hunk 19]
```diff
@@ -14,9 +14,79 @@
...
+    job = estimator.run([(psi1, hamiltonian1, [theta1])])
+    job_result = job.result() # It will block until the job finishes.
+    print(f"The primitive-job finished with result {job_result}"))
+
+    # calculate [ [<psi1(theta1)|H1|psi1(theta1)>,
+    #              <psi1(theta3)|H3|psi1(theta3)>],
+    #             [<psi2(theta2)|H2|psi2(theta2)>] ]
+    job2 = estimator.run(
+        [(psi1, [hamiltonian1, hamiltonian3], [theta1, theta3]), (psi2, hamiltonian2, theta2)]
+    )
```
Comment (ihincks) [inline qiskit/primitives/base/base_estimator.py:82]:
```suggestion
    # calculate [ [<psi1(theta1)|H1|psi1(theta1)>,
    #              <psi1(theta3)|H3|psi1(theta3)>],
    #             [<psi2(theta2)|H2|psi2(theta2)>] ]
    # aiming for precision to 0.02 on all estimates
    job2 = estimator.run(
        [(psi1, [hamiltonian1, hamiltonian3], [theta1, theta3]), (psi2, hamiltonian2, theta2)],
        precision=0.02
    )
```

File: qiskit/primitives/base/base_estimator.py (hunk 15 above)
Comment (ihincks) [reply qiskit/primitives/base/base_estimator.py]: The priority is:
 1. whatever's in a pub
 2. if that is none, whatever's in the run kwarg
 3. if that is none, the primitive chooses some value, which can be implementation specific.

File: qiskit/primitives/base/base_estimator.py [hunk 20]
```diff
@@ -254,3 +328,41 @@ def parameters(self) -> tuple[ParameterView, ...]:
...
+    def run(self, pubs: Iterable[EstimatorPubLike], precision: float | None = None) -> Job:
+        """Estimate expectation values for each provided pub (Primitive Unified Bloc).
+
+        Args:
+            pubs: An iterable of pub-like objects, such as tuples ``(circuit, observables)`` or
+                  ``(circuit, observables, parameter_values)``.
+            precision: The target precision for expectation value estimates of each
+                       run :class:`.EstimatorPub` that does not specify its own
+                       precision. If None the estimator's default precision value
+                       will be used.
```
Comment (ihincks) [inline qiskit/primitives/base/base_estimator.py:375]:
```suggestion
            precision: The target precision for expectation value estimates of each
                       pub that does not specify its own precision. If ``None`` the 
                       estimator's default precision value will be used.
```

File: qiskit/primitives/containers/estimator_pub.py [hunk 21]
```diff
@@ -0,0 +1,168 @@
...
+    """Primitive Unified Bloc for any Estimator primitive.
+
+    An estimator pub is essentially a tuple ``(circuit, observables, parameter_values, precision)``.
+
+    If precision is provided this should be used for the target precision of an
+    estimator, if ``precision=None`` the estimator will determine the target precision.
+    """
+
+    __slots__ = ("_circuit", "_observables", "_parameter_values", "_precision", "_shape")
+
```
Comment (ihincks) [inline qiskit/primitives/containers/estimator_pub.py:42]: Using slots seems unnecessary
```suggestion
```

File: qiskit/primitives/containers/estimator_pub.py [hunk 22]
```diff
@@ -0,0 +1,168 @@
...
+        precision: float | None = None,
+        validate: bool = False,
+    ):
+        """Initialize an estimator pub.
+
+        Args:
+            circuit: A quantum circuit.
+            observables: An observables array.
+            parameter_values: A bindings array, if the circuit is parametric.
+            precision: An optional target precision for expectation value estimates.
```
Comment (ihincks) [inline qiskit/primitives/containers/estimator_pub.py:57]:
```suggestion
            precision: An optional target precision for expectation value estimates.
                This value takes precedence over any value owed by or supplied to an
                estimator.
```

File: qiskit/primitives/containers/estimator_pub.py [hunk 23]
```diff
@@ -0,0 +1,168 @@
...
+        return self._observables
+
+    @property
+    def parameter_values(self) -> BindingsArray:
+        """A bindings array."""
+        return self._parameter_values
+
+    @property
+    def precision(self) -> float | None:
+        """The target precision for expectation value estimates (optional)."""
```
Comment (ihincks) [inline qiskit/primitives/containers/estimator_pub.py:100]:
```suggestion
        """The target precision for expectation value estimates (optional).
        
        This value takes precedence over any value owed by or supplied to an estimator.
        """
```

File: qiskit/primitives/containers/estimator_pub.py [hunk 24]
```diff
@@ -0,0 +1,168 @@
...
+        return self._precision
+
+    @classmethod
+    def coerce(cls, pub: EstimatorPubLike, precision: float | None = None) -> EstimatorPub:
+        """Coerce :class:`~.EstimatorPubLike` into :class:`~.EstimatorPub`.
+
+        Args:
+            pub: A compatible object for coercion.
+            precision: an optional default precision to use if not
+                       already specified by the pub-like object.
```
Comment (ihincks) [inline qiskit/primitives/containers/estimator_pub.py:110]:
```suggestion
            precision: An optional default precision to use if not
                       already specified by the pub-like object.
```

**Review by ihincks — COMMENTED** [review]
Looks like we'll need some tests for `EstimatorPub` too.

File: qiskit/primitives/base/base_estimator.py (hunk 14 above)
Comment (ihincks) [reply qiskit/primitives/base/base_estimator.py]: In the latest round, this arg has been removed from the constructor and added to the run method and to the pub attributes. Ihe value in the pub, if present, takes precedence. The value in the run kwargs, if present sets the value for each pub where it is not set. The primitive must choose a value for precision if none are given, and this is left up to each implementation to decide. For example, state vector simulators have no reason to have precision >0 (except, of course, machine epsilon, which we can ignore).

File: qiskit/primitives/containers/observables_array.py [hunk 25]
```diff
@@ -211,7 +211,7 @@ def coerce(cls, observables: ObservablesArrayLike) -> ObservablesArray:
     def validate(self):
         """Validate the consistency in observables array."""
         num_qubits = None
-        for obs in self._array:
+        for obs in self._array.reshape(-1):
```
Comment (t-imamichi) [inline qiskit/primitives/containers/observables_array.py:214]: I'm not sure why this change is included. I guess some validations may fail due to the previous version. It might be good to add a unit test of observable array.

File: qiskit/primitives/base/base_estimator.py [hunk 26]
```diff
@@ -254,3 +336,43 @@ def parameters(self) -> tuple[ParameterView, ...]:
...
+    An Estimator implementation must treat the :meth:`.run` method ``precision=None``
+    kwarg as using a default ``precision`` value.  The default value and methods to
+    set it can be determined by the Estimator implementor.
+    """
+
+    @staticmethod
+    def _make_data_bin(pub: EstimatorPub) -> DataBin:
+        # provide a standard way to construct estimator databins to ensure that names match
+        # across implementations
+        return make_data_bin((("evs", NDArray[np.float]), ("stds", NDArray[np.float])), pub.shape)
```
Comment (ikkoham) [inline qiskit/primitives/base/base_estimator.py]: Maybe `np.float64` might be better?

Comment (t-imamichi) [reply qiskit/primitives/base/base_estimator.py]: I think `float` is better since `np.float` was a deprecated alias for the builtin `float`.

> `np.float` was a deprecated alias for the builtin `float`. To avoid this error in existing code, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
The aliases was originally deprecated in NumPy 1.20; for more details and guidance see the original release note at:
    https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations

File: qiskit/primitives/base/base_estimator.py [hunk 27]
```diff
@@ -254,3 +336,43 @@ def parameters(self) -> tuple[ParameterView, ...]:
...
+
+    @abstractmethod
+    def run(
+        self, pubs: Iterable[EstimatorPubLike], precision: float | None = None
```
Comment (t-imamichi) [inline qiskit/primitives/base/base_estimator.py]: It might be safer to force `precision` keyword argument.
```suggestion
        self, pubs: Iterable[EstimatorPubLike], *, precision: float | None = None
```

File: qiskit/primitives/containers/observables_array.py (hunk 25 above)
Comment (ihincks) [reply qiskit/primitives/containers/observables_array.py:214]: Yes, it was given an error about not being iterable when `shape=()`, and it was also not working properly when ndim>1. This reshape trick gurantees it to always be a vector. But you're right that I should have put a test in.

File: qiskit/primitives/base/base_estimator.py (hunk 27 above)
Comment (chriseclectic) [reply qiskit/primitives/base/base_estimator.py]: Sure, we should add to sampler for shots kwarg as well to be consistent in that case

File: qiskit/primitives/base/base_estimator.py (hunk 26 above)
Comment (chriseclectic) [reply qiskit/primitives/base/base_estimator.py]: I changed to `np.float64` here

File: qiskit/primitives/containers/observables_array.py (hunk 25 above)
Comment (ihincks) [reply qiskit/primitives/containers/observables_array.py:214]: A test was added in 12c71e5

**Review by ihincks — APPROVED** [review]
LGTM. Thanks all.
